lib/backend/k8s/k8s.go

```diff
@@ -173,33 +175,46 @@ func (c *KubeClient) ensureThirdPartyResources() error {
 // createThirdPartyResources creates the necessary third party resources if they
 // do not already exist.
 func (c *KubeClient) createThirdPartyResources() error {
+	// We can check registration of the different custom resources in
+	// parallel.
+	done := make(chan error)
+
 	// Ensure a resource exists for Calico global configuration.
-	log.Info("Ensuring GlobalConfig ThirdPartyResource exists")
-	tpr := extensions.ThirdPartyResource{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      "global-config.projectcalico.org",
-			Namespace: "kube-system",
-		},
-		Description: "Calico Global Configuration",
-		Versions:    []extensions.APIVersion{{Name: "v1"}},
-	}
-	_, err := c.clientSet.Extensions().ThirdPartyResources().Create(&tpr)
-	if err != nil {
-		// Don't care if it already exists.
-		if !kerrors.IsAlreadyExists(err) {
-			return resources.K8sErrorToCalico(err, tpr)
+	go func() {
+		log.Info("Ensuring GlobalConfig ThirdPartyResource exists")
+		tpr := extensions.ThirdPartyResource{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "global-config.projectcalico.org",
+				Namespace: "kube-system",
+			},
+			Description: "Calico Global Configuration",
+			Versions:    []extensions.APIVersion{{Name: "v1"}},
+		}
+		_, err := c.clientSet.Extensions().ThirdPartyResources().Create(&tpr)
+		if err != nil {
+			// Don't care if it already exists.
+			if !kerrors.IsAlreadyExists(err) {
+				done <- resources.K8sErrorToCalico(err, tpr)
+				return
+			}
+		}
+		done <- nil
+		return
+	}()
+
+	go func() { done <- c.ipPoolClient.EnsureInitialized() }()
+	go func() { done <- c.snpClient.EnsureInitialized() }()
+	go func() { done <- c.globalBgpClient.EnsureInitialized() }()
+
+	// Wait for all 4 registrations to complete and keep track of the last
+	// error to return.
+	var lastErr error
+	for i := 0; i < 4; i++ {
```

Comment:
I did look at that, but you still need to update the count of the WaitGroup so that it waits for the correct number.

Longer term, I'd like these things to be registered and therefore the counting would be automated and not an extra bit of code that needs adding.  But I can use a WG if it's irking you.